driver-core/src/main/com/mongodb/internal/connection/IdHoldingBsonWriter.java

```diff
@@ -414,23 +414,27 @@ private void setCurrentFieldName(final String name) {
     }
 
     private boolean isWritingId() {
-        return getIdBsonWriter().getCurrentLevel() >= 0 || (getCurrentLevel() == 0 && currentFieldName != null
+        return getIdBsonWriterCurrentLevel() >= 0 || (getCurrentLevel() == 0 && currentFieldName != null
                 && currentFieldName.equals(ID_FIELD_NAME));
     }
 
     private void addBsonValue(final Supplier<BsonValue> value, final Runnable writeValue) {
         if (isWritingId()) {
-            if (getIdBsonWriter().getCurrentLevel() >= 0) {
+            if (getIdBsonWriterCurrentLevel() >= 0) {
                 writeValue.run();
             } else {
                 id = value.get();
             }
         }
     }
 
+    private int getIdBsonWriterCurrentLevel() {
+        return idBsonBinaryWriter == null ? -1 : idBsonBinaryWriter.getCurrentLevel();
+    }
+
     private LevelCountingBsonWriter getIdBsonWriter() {
         if (idBsonBinaryWriter == null) {
-            outputBuffer = new BasicOutputBuffer();
+            outputBuffer = new BasicOutputBuffer(128);
```

Comment:
This has the potential to affect performance both positively and negatively. But I agree that 128 bytes is likely a better estimation for the size of the `_id` value than 1024 bytes.